apps/gdalalg_vector_update.cpp

```diff
@@ -0,0 +1,350 @@
+
+#include "gdalalg_vector_update.h"
+
+#include "ogr_p.h"
+#include "ogrsf_frmts.h"
+
+//! @cond Doxygen_Suppress
+
+#ifndef _
+#define _(x) (x)
+#endif
+
+/************************************************************************/
+/*        GDALVectorUpdateAlgorithm::GDALVectorUpdateAlgorithm()        */
+/************************************************************************/
+
+GDALVectorUpdateAlgorithm::GDALVectorUpdateAlgorithm(bool standaloneStep)
+    : GDALVectorPipelineStepAlgorithm(NAME, DESCRIPTION, HELP_URL,
+                                      ConstructorOptions()
+                                          .SetStandaloneStep(standaloneStep)
+                                          .SetInputDatasetMaxCount(1)
+                                          .SetAddInputLayerNameArgument(false)
+                                          .SetAddDefaultArguments(false))
+{
+    if (standaloneStep)
+    {
+        AddVectorInputArgs(false);
+    }
+    {
+        auto &layerArg = AddArg(GDAL_ARG_NAME_INPUT_LAYER, 0,
+                                _("Input layer name"), &m_inputLayerNames)
+                             .SetMaxCount(1);
+        auto inputArg = GetArg(GDAL_ARG_NAME_INPUT);
+        if (inputArg)
+            SetAutoCompleteFunctionForLayerName(layerArg, *inputArg);
+    }
+
+    AddProgressArg();
+
+    AddOutputDatasetArg(&m_outputDataset, GDAL_OF_VECTOR)
+        .SetDatasetInputFlags(GADV_NAME | GADV_OBJECT);
+    AddOutputOpenOptionsArg(&m_outputOpenOptions);
+    AddArg(GDAL_ARG_NAME_OUTPUT_LAYER, 0, _("Output layer name"),
+           &m_outputLayerName);
+
+    m_update = true;
+    AddUpdateArg(&m_update).SetDefault(true).SetHidden();
+
+    AddArg("mode", 0, _("Set update mode"), &m_mode)
+        .SetDefault(m_mode)
+        .SetChoices(MODE_MERGE, MODE_UPDATE_ONLY, MODE_APPEND_ONLY);
+
+    AddArg("key", 0, _("Field(s) used as a key to identify features"), &m_key)
+        .SetPackedValuesAllowed(false);
+}
+
+/************************************************************************/
+/*                  GDALVectorUpdateAlgorithm::RunStep()                */
+/************************************************************************/
+
+bool GDALVectorUpdateAlgorithm::RunStep(GDALPipelineStepRunContext &ctxt)
+{
+    auto poSrcDS = m_inputDataset[0].GetDatasetRef();
+    CPLAssert(poSrcDS);
+
+    auto poDstDS = m_outputDataset.GetDatasetRef();
+    CPLAssert(poDstDS);
+    CPLAssert(poDstDS->GetAccess() == GA_Update);
+
+    auto poSrcDriver = poSrcDS->GetDriver();
+    auto poDstDriver = poDstDS->GetDriver();
+    if (poSrcDS == poDstDS ||
+        (poSrcDriver && poDstDriver &&
+         !EQUAL(poSrcDriver->GetDescription(), "MEM") &&
+         !EQUAL(poDstDriver->GetDescription(), "MEM") &&
+         strcmp(poSrcDS->GetDescription(), poDstDS->GetDescription()) == 0))
+    {
+        ReportError(CE_Failure, CPLE_NotSupported,
+                    "Source and destination datasets must be different");
```

Comment:
Source and target? Input and output? Origin and destination? 
We have to try to be consistent.